docs/architecture/how/resharding.md

```diff
@@ -0,0 +1,107 @@
+
+**TODO**: It seems that the code was written with the idea of persisting the "progress" (that's why have parts etc) - but AFAIK currently we never persist the progress of the resharding - so if the node restarts, we're starting from scratch.
+
+
+Extracting of the account from the key happens in ``parse_account_id_from_raw_key`` - and we do it for all types of data that we store in the trie (contract code, keys, account info etc) EXCEPT for Delayed receipts. Then, we figure out the shard that this account is going to belong to, and we add this key/value to that new trie.
+
+This way, after going over all the key/values from the original trie, we end up with X new tries (one for each new shard).
+
+
+IMPORTANT: in the current code, we only support such 'splitting' (so a new shard can have just one parent).
+
+**Discussion point**: When should we start supporting the 'merge' operation ? And when the combination (split & merge)?
+
+
+### Why delayed receipts are special?
+For all the other columns, there is no dependency between entries, but in case of delayed receipts - we are forming a 'queue'. We store the information about the first index and the last index (in DelayedReceiptIndices struct).
+
+Then, we receipt arrives, we add it as the 'DELATED_RECEIPT + last_index' key (and increment last_index by 1).
+
+That's why we cannot move it column 'in trie parts' (like we did for others), but we do it by 'iterating' over this queue.
+
+
+
+## What should we improve to achieve 'stressless' resharding?
+
+Currently trie is split sequencially (shard by shard), and also sequancially within a shard - by iterating over all the elements. 
+
+This process must finish within a single epoch - which is might be challenging (especially for larger archival nodes).
+
+### Each shard should be handled independently
+This is a low-hanging fruit - but each shard's splitting should be done in a separate thread.
+
+
+### Do trie splitting on the level of accounts, not key/values
+
+Most of the keys in our tries, are in the form of ``${some short prefix}${account_id}${long suffix}``, we could use this fact, and move the whole 'trie subtrees' around.
+
+In theory, this should be super-fast - as the hash of such tree doesn't change (so all the trie nodes below have the same hashes etc).
+
+Unfortunately we add the ShardUid prefix to the trie keys, when storing them in database - which means, that we'd have to iterate over all of them to copy the values to a new row (with a new sharduid) - which kills all the benefits.
+
+#### should we remove the sharduid prefix?
+
+one advantage of sharduid, is that it allows us to quickly remove the contents of the shard (which we do when we're about to do statesync - as we want to start with a clear slate).
```

Comment:
The reason that shardid/sharduid is used as prefix for the storage is because of https://github.com/near/nearcore/pull/2638. TLDR: if shardid is not a prefix, then different shards can share trie nodes, and that could lead problems in gc with the current implementation, so a long time ago, we fixed the issue by storing tries from different shards separately. 

There are probably ways to fix the refcount issues without separating tries from different shards, for example, merging the TrieUpdates in a DBTransactions before committing them.